package-builder/flox-build.mk

```
@@ -157,6 +157,10 @@ define BUILD_local_template =
 	@echo "Building $(_name) in local mode"
 	$(if $(_virtualSandbox),$(PRELOAD_ARGS) FLOX_SRC_DIR=$$$$($(_pwd)) FLOX_VIRTUAL_SANDBOX=$(_sandbox)) \
 	MAKEFLAGS= out=$(_out) $(FLOX_ENV)/activate --turbo -- $(_bash) -e $($(_pvarname)_buildScript)
+	@if [ ! -e "$(_out)" ]; then \
+		echo 'ERROR: build did not produce expected $$$$out' 1>&2; \
```

It seems the reference to `$$$$out` is an attempt to report the `$out` environment variable at runtime, but I don't believe that will be defined at this point.

Also, should we pattern this more closely after the equivalent Nix error?

Combining these two observations, this would be the closest approximation I can imagine: 

```suggestion
    echo "ERROR: builder failed to produce output path for output '$(_pname)' at '$(_out)'" 1>&2; \
```

FTR, the Nix equivalent is:
```
error: builder for '/nix/store/n6lq8chmhb3m66f06ndi7i2masyi1kvg-foo-0.0.0.drv' failed to produce output path for output 'out' at '/nix/store/sdn5wkzzddh14ghzcb96fjk6n2pipm50-foo-0.0.0'
```

---

Briefly met with Graham to discuss how we should deliver consistent messages across the various sandbox modes, and we landed on the following:

* let the Nix build put out whatever message it wants to provide
* regardless, we should have a consistent Flox-curated error message for the failure to populate `$out`
    * in the `sandbox=(off|warn|enforce)` modes we report that from line 160 above
    * in the `sandbox=(pure)` mode we will need to grep for the "failed to produce output path for output" message in the build logs and then report _the exact same message_ as above
* finally, we report a "Hint" that gives the user an idea of what to do to solve the problem

So ... an example local build could be:
```
❌ ERROR: Build command is missing outputs copied to $out.
Hint: populate the '$out/bin' directory from within <buildname>.command, e.g. with make install PREFIX=$out
```

An an example pure build could be:
```
error: builder for '/nix/store/n6lq8chmhb3m66f06ndi7i2masyi1kvg-foo-0.0.0.drv' failed to produce output path for output 'out' at '/nix/store/sdn5wkzzddh14ghzcb96fjk6n2pipm50-foo-0.0.0'
❌ ERROR: Build command is missing outputs copied to $out.
Hint: populate the '$out/bin' directory from within <buildname>.command, e.g. with make install PREFIX=$out
```

---

> It seems the reference to `$$$$out` is an attempt to report the `$out` environment variable at runtime, but I don't believe that will be defined at this point.

My intention was to report a literal `$out` without any interpolation because it has a special user-facing meaning. We don't want the user to write directly to the temporary directory that `$out` evaluates to because it won't be the same on subsequent builds.

> * let the Nix build put out whatever message it wants to provide
> * regardless, we should have a consistent Flox-curated error message for the failure to populate `$out`
>   * …
>   * in the `sandbox=(pure)` mode we will need to grep for the "failed to produce output path for output" message in the build logs and then report _the exact same message_ as above

Can you describe why you'd approach it this way? Is it based on the understanding that the Nix error will always appear first? (see other thread below)

> Of course the colour of this particular bikeshed (i.e. the exact wording) shall evolve perpetually hereafter, but as long as we create the structure for reporting the consistent error message and hint in the right places then we'll be in good shape. 👍

We can tweak the messaging and hint in this PR but I would like to defer the structure for error reporting to another ticket because it's a lot more work that would bleed over into the next sprint and delay us from getting this fix in for user testing.

Currently we have a stream of unstructured STDOUT/STDERR strings and a status code available to us on the Rust side, much like the `pkgdb` integration, and we need to make some decisions about how and how much of that we interpret/display/test. That prevents us from providing a more granular final error than the current `❌ ERROR: Build failed with status: exit status: 2`.

> finally, we report a "Hint" that gives the user an idea of what to do to solve the problem

I guess we should account for them not using `make install` (like the original report in #2091) or outputting a single file instead of a directory. How about the following?

```
ERROR: Build command did not copy outputs to '$out'.
- copy a single file with 'cp bin $out'.
- copy multiple files with 'mkdir -p $out && cp bin/* $out/'
- copy files from an Autotools project with 'make install PREFIX=$out'
```

---

As dan said, currently we  only have the raw make output available in rust, so the best error message we could give, is "Some of your builds failed because of $out not being populated. Which one? 🤷🏼‍♂️."
I think the best position to do any of this error detection is likely the makefile itself where all relevant information is available.
The make file can format and print the error to stderr before it terminates.
This still does not quite eliminate all the other blurb that make might print afterwards, or the fact that the cli only knows that the make process finished unsuccessfully.

So to inform the cli of what occured, alternatively it could write errors structured or formatted to a separate channel which we read on the rust side to pass on the error.
Unfortunately, std* are both in use for build logs.. so it would rather be a separate pipe that we point to with an argument.

---

I guess this logging stuff can get arbitrarily complex, but providing a diagnostic/control channel seems somewhat reasonable considering that std* can have completely arbitrary output

---

I've rebased with the updated message and will create two new tickets (probably tomorrow) for output design and "function logging".

---

@dcarley might overlap with https://github.com/flox/flox/issues/2044 as well

---

Ahh thanks! I'll add notes to that instead and create a separate issue for "calling function hook".